Automattic/wp-calypso
# A4A > Referrals: Implement Client Checkout

Closes https://github.com/Automattic/jetpack-genesis/issues/372

## Proposed Changes

This PR implements client checkout

NOTE: 

- [Design](https://www.figma.com/design/fuufP6VNfZXYmvLsTRWRlG/Referrals?node-id=7600-30663&m=dev)
- **IMPORTANT**: **Please don't forget to do regression testing by logging into the checkout and payment methods page as an agency.**


## Testing Instructions

1. Checkout to the branch locally and start the server.
2. Create a new WPCOM account, but don't sign up for A4A if you have not tested any PR related to the client before.
3. Get the ID(WPCOM ID) from this API that gets called when you log in: 
https://public-api.wordpress.com/rest/v1.1/me and note it down.
4. Login to your agency account > Go to Marketplace > Toggle Refer products > Select product > Go to Checkout > Enter the email ID of the newly created account > Open the network tab > Click the `Request payment from client` button > Get the ID (REFERRAL_ID) from the API response that gets called when the button is clicked.
6. Logout and log in to your client account 
7. Go to /client/payment-methods and delete all the payment methods.
8. Follow the instructions here to get the secret: 34c3e-pb
9. Go to /client/checkout?agency_id={id}&referral_id={id}&secret={secret}
10. Verify that you see the Checkout UI as shown below & verify that the products listed are correct.

<img width="1728" alt="Screenshot 2024-06-13 at 11 28 43 AM" src="https://github.com/Automattic/wp-calypso/assets/10586875/58c676ce-efeb-4c69-b9d3-c16587393022">

11. Click the `Add my payment method` button > Verify that you are directed to the add payment page as shown below:

<img width="1728" alt="Screenshot 2024-06-14 at 3 19 06 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/7ae2b737-614d-4a6b-a7f9-4fcc5de4e70e">


12. Add a new card & verify that you are redirected to the checkout and it looks like below:

<img width="1728" alt="Screenshot 2024-06-14 at 2 41 57 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/2e952ac1-eab2-4439-bcb8-0880b6e6193c">

13. Click the `Submit my payment information` and verify that you get a success message and you are redirected to /client/subscriptions.
14. Go to /client/checkout?agency_id={id}&referral_id={id}&secret={secret} again and try submitting it again for the same referral and verify you get the appropriate error message and you are redirected to the subscriptions page.

<img width="1728" alt="Screenshot 2024-06-14 at 6 31 35 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/fcfaf2d0-ef1d-483b-bda7-d6559cc5ec0e">


Other Error/success notices.

<img width="624" alt="Screenshot 2024-06-12 at 5 56 20 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/c359d932-27d8-43ed-b57d-e68dd5574d79">
<img width="596" alt="Screenshot 2024-06-12 at 6 03 00 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/5231497f-955b-4c6e-a382-087586cfe765">
<img width="471" alt="Screenshot 2024-06-12 at 6 04 26 PM" src="https://github.com/Automattic/wp-calypso/assets/10586875/47686abd-8ac4-490f-b5ac-aa67d489a964">

15. Please remember to perform regression testing by logging in as an agency on the checkout and payment methods page.

## Pre-merge Checklist
